File: verifier-core/src/main/java/com/jetbrains/pluginverifier/verifiers/field/FieldTypeVerifier.java
package com.jetbrains.pluginverifier.verifiers.field;

import com.intellij.structure.resolvers.Resolver;
import com.jetbrains.pluginverifier.api.VContext;
import com.jetbrains.pluginverifier.location.ProblemLocation;
import com.jetbrains.pluginverifier.problems.ClassNotFoundProblem;
import com.jetbrains.pluginverifier.utils.VerifierUtil;
import org.objectweb.asm.tree.ClassNode;
import org.objectweb.asm.tree.FieldNode;

/**
 * @author Dennis.Ushakov
 */
public class FieldTypeVerifier implements FieldVerifier {
  public void verify(final ClassNode clazz, final FieldNode field, final Resolver resolver, final VContext ctx) {
    final String className = VerifierUtil.extractClassNameFromDescr(field.desc);

    if (className == null || VerifierUtil.classExistsOrExternal(ctx, resolver, className)) {
      return;
    }

    ctx.registerProblem(new ClassNotFoundProblem(className), ProblemLocation.fromField(clazz.name, field.name));
  }
}

File: verifier-core/src/main/java/com/jetbrains/pluginverifier/verifiers/field/FieldTypeVerifier.kt
package com.jetbrains.pluginverifier.verifiers.field

import com.jetbrains.pluginverifier.verifiers.VerificationContext
import com.jetbrains.pluginverifier.verifiers.createFieldLocation
import com.jetbrains.pluginverifier.verifiers.extractClassNameFromDescr
import com.jetbrains.pluginverifier.verifiers.resolveClassOrProblem
import org.objectweb.asm.tree.ClassNode
import org.objectweb.asm.tree.FieldNode

/**
 * @author Dennis.Ushakov
 */
class FieldTypeVerifier : FieldVerifier {
  override fun verify(clazz: ClassNode, field: FieldNode, ctx: VerificationContext) {
    val className = field.desc.extractClassNameFromDescr() ?: return

    ctx.resolveClassOrProblem(className, clazz) { createFieldLocation(clazz, field) }
  }
}
